src/Persistence/PersistenceManager.php

```diff
@@ -165,26 +172,45 @@ public function refresh(object &$object, bool $force = false): object
 
         $om = $strategy->objectManagerFor($object::class);
 
-        if ($strategy->contains($object)) {
-            try {
-                $om->refresh($object);
-            } catch (\LogicException|\Error) {
-                // prevent entities/documents with readonly properties to create an error
-                // LogicException is for ORM / Error is for ODM
-                // @see https://github.com/doctrine/orm/issues/9505
-            }
-
-            return $object;
-        }
-
         if ($strategy->isEmbeddable($object)) {
             return $object;
         }
 
-        $id = $om->getClassMetadata($object::class)->getIdentifierValues($object);
+        if (!$strategy->contains($object)) {
+            $objectFromDb = null;
+
+            $id = $om->getClassMetadata($object::class)->getIdentifierValues($object);
+
+            if ($id) {
+                // "merge" object if it is not managed
+                $objectFromDb = $om->find($object::class, $id);
+            }
 
-        if (!$id || !($object = $om->find($object::class, $id))) { // @phpstan-ignore parameterByRef.type
-            throw RefreshObjectFailed::objectNoLongExists();
+            if ($objectFromDb) {
+                $object = $objectFromDb;
+            } else {
+                if ($allowRefreshDeletedObject) {
+                    $identifierFields = $om->getClassMetadata($object::class)->getIdentifierFieldNames();
+                    foreach ($identifierFields as $field) {
+                        try {
+                            set($object, $field, null);
```

Comment:
@kbond any thoughts about this?

I'm resetting to null ids that can be set null (usually, it is the ones that are generated by the DB I guess)

Reply:
What would be the problem with doing away with this logic and always throwing an `objectNotLongerExists()` exception?

Reply:
this test would break badly :sweat_smile: 

```php
    public function test_remove_object(): void
    {
        $object = $this->factory()->create();

        $client = self::createClient();
        $client->request('GET', "/delete/{$object->id}");
        self::assertResponseIsSuccessful();

        // here, $object is a lazy proxy, so any access to any method/prop will throw `RefreshObjectFailed`
    }
```

Reply:
solved now: I've reverted reset of the ids, and implemented functions `assertPersisted()` / `assertNotPersisted()`